rekall-core/rekall/plugins/addrspaces/vmi.py

```diff
@@ -24,21 +36,30 @@ def __init__(self, base=None, filename=None, session=None, **kwargs):
         url = filename or (session and session.GetParameter("filename"))
         self.as_assert(url, "Filename must be specified in session (e.g. "
                        "session.SetParameter('filename', 'vmi://domain').")
-        self.as_assert(url.startswith(URL_PREFIX),
-                       "The domain must be passed with the URL prefix {}".format(URL_PREFIX))
-        domain = url[len(URL_PREFIX):]
-        self.as_assert(domain, "domain name missing after {}".format(URL_PREFIX))
+        pattern = '{}((?P<hypervisor>xen|kvm)/)?(?P<domain>.*)'.format(URL_PREFIX)
+        m = re.match(pattern, url)
+        self.as_assert(m,
+                "VMI URL format: {}(xen|kvm)?<domain>".format(URL_PREFIX))
+        domain = m.group('domain')
+        hypervisor = m.group('hypervisor')
+        self.as_assert(domain, "no domain name specified")
+        self.mode = None
+        if hypervisor:
+            self.mode = VMIMode[hypervisor.upper()]
 
         super(VMIAddressSpace, self).__init__(base=base, session=session, **kwargs)
-        self.vmi = Libvmi(domain, partial=True)
+        self.vmi = Libvmi(domain, mode=self.mode, partial=True)
+        self.min_addr = 0
+        self.max_addr = self.vmi.get_memsize() - 1
 
     def close(self):
         self.vmi.destroy()
 
     def read(self, addr, size):
-        buffer, bytes_read = self.vmi.read_pa(addr, size)
-        if bytes_read != size:
-            raise RuntimeError('Error while reading physical memory at {}'.format(hex(addr)))
+        try:
+            buffer, bytes_read = self.vmi.read_pa(addr, size)
```

Comment:
It is up to you but if you already know the read will fall outside the allowed range you might want to avoid calling into libvmi because it cause it to log errors anyway.

I am not sure how the underlying library runs the reads but maybe its worth breaking the reads up into parts which fall within allowable ranges and those which should be zero filled. The easiest way to do this is to overlay a run based address space on top of the raw libvmi address space and then program the run based address space with the allowed runs. Can you get the list of valid runs from libvmi or is it capable of reading everywhere and null padding by itself?

Reply:
I'm calling @tklengyel here since he knows better libvmi internals =)

Reply:
LibVMI doesn't pad reads to match what the requested size was. If you are reading such that it spans multiple pages, and one of those pages is unmappable, it will error out and return to you as much as it was able to read and tell you what that was in `bytes_read`, see https://github.com/libvmi/libvmi/blob/master/libvmi/read.c#L118. If you need a padded read, you can implement a wrapper around this to do so.

Reply:
I could implement this padded read directly in the Libvmi python bindings instead of copying the definition in both Rekall and Volatility as.

Reply:
Is there a way to get the mapped pages from libvmi in advance? Otherwise you could just break up the reads into separate page size reads and issue a single call per page - then if the read fails you just pad with a single page of zeros for that read and keep trying the rest of the range.

This would slow down large scanning plugins (which typically read 10mb at a time) but probably wont make too much difference for plugins which read a few bytes at a time.

Reply:
>It is up to you but if you already know the read will fall outside the allowed range you might want to avoid calling into libvmi because it cause it to log errors anyway.
I would prefer that also, because libvmi will generate errors on stderr:
~~~
VMI_ERROR: --requesting PA [0xff001000] beyond max physical address [0xff000000]
VMI_ERROR: 	paddr: ff000000, length 1000, vmi->max_physical_address ff000000
VMI_ERROR: create_new_entry failed
~~~

Why does Rekall tries to read outside of the given mappings that we have defined in the AddressSpace with `get_mappings()/is_valid_address()/get_available_addresses` ?
When these methods are used inside Rekall ?

> Is there a way to get the mapped pages from libvmi in advance?
You can get the mapped pages of a specific process:
https://github.com/libvmi/libvmi/blob/master/libvmi/libvmi_extra.h#L46
Not sure that's what you were asking by mapped pages ?

Reply:
I was more referring to the physical memory pages which are backed by actual RAM. Not all of physical memory is readable and there are only some valid mapped regions which are backed by memory (e.g. the output of /proc/iomem - the areas marked as System RAM).

Rekall should not be reading outside these regions normally but this appears to be what is happening here. Can you verify that get_mappings() returns the correct regions - the error message above seems to indicate that the maximum physical address should be 0xff000000. 

I guess this is a good point - we should answer why Rekall is trying to access outside these regions in the first place. I suspect that there is some inconsistency between the return of get_mappings() and the allowed physical address regions.

Reply:
>we should answer why Rekall is trying to access outside these regions in the first place. I suspect that there is some inconsistency between the return of get_mappings() and the allowed physical address regions.

When rekall calls the read method, is there a filter (like `is_valid_address` beforehand to check if the requested address is in the range that we defined ?

If so, does it check if `addr` is in the range, or `addr` + `count` ?

Any idea how i can start to get some information ?